### PR TITLE
feat: directUrl support

### PR DESCRIPTION
**Pull request by eviefp**
Add support for `directUrl`.

This is a follow-up for https://github.com/prisma/prisma-engines/pull/3547 and https://github.com/prisma/prisma-engines/pull/3551

I followed the usages of `getConfig`, which is where these changes sprung up from. I would love it if someone would do the same to validate I did not miss anything.

Also, I am unsure on how to test this. Happy to be given some direction here.

Part of https://github.com/prisma/prisma/issues/6485

### CODE REVIEW

**Comment by Jolg42** [comment]
@eviefp Could you sign the CLA from https://github.com/prisma/prisma/pull/17210#issuecomment-1375942082?

Also, it looks like we are missing tests for the error message 👀

**Comment by Jolg42** [comment]
We could add a test, for
- https://github.com/prisma/prisma/blob/main/packages/migrate/src/__tests__/ensureDatabaseExists.test.ts
- https://github.com/prisma/prisma/blob/main/packages/migrate/src/__tests__/MigrateDeploy.test.ts

Not sure if it makes sense to test all commands using this for this error message, maybe one is enough.

**Comment by eviefp** [comment]
Added a test for reset, would that be enough?

File: packages/internals/src/engine-commands/getConfig.ts [hunk 1]
```diff
@@ -83,6 +83,12 @@ ${detailsHeader} ${message}`
   }
 }
 
+export function getDirectUrl(ds: DataSource): EnvValue {
```
Comment (janpio) [inline packages/internals/src/engine-commands/getConfig.ts:86]: Nit:

This confused me a bit: Why is the method that 99% returns the `url` and only sometimes the `directUrl` called `getDirectUrl`?

But I think I understand now: Either `url` or `directUrl` are the _direct URL_ we want to use. Is that your motivation for this naming?

As "direct URL" and "`directUrl`" are the same term, I think I might name the function `getEffectiveUrl` or something like that.

Comment (Jolg42) [reply packages/internals/src/engine-commands/getConfig.ts:86]: A rename would be nice to avoid confusion, can be done later if needed

**Review by Jolg42 — APPROVED** [review]
Looks good! Let's merge this :shipit: 

About the test failure, a snapshot update is needed or maybe you forgot to add the dev.db file to git?

(see comments)

File: packages/internals/src/cli/checkUnsupportedDataProxy.ts [hunk 2]
```diff
@@ -62,8 +62,9 @@ async function checkUnsupportedDataProxyMessage(command: string, args: Args, imp
 
       const datamodel = await fs.promises.readFile(argValue, 'utf-8')
       const config = await getConfig({ datamodel, ignoreEnvVarErrors: true })
-      const urlFromValue = config.datasources[0]?.url.value
-      const urlEnvVarName = config.datasources[0]?.url.fromEnvVar
+      const url = getEffectiveUrl(config.datasources[0])
+      const urlFromValue = url.value
+      const urlEnvVarName = url.fromEnvVar
```
Comment (jkomyno) [inline packages/internals/src/cli/checkUnsupportedDataProxy.ts:67]: Nit:

```suggestion
      const { value: urlFromValue, fromEnvVar: urlFromEnvVarName } = getEffectiveUrl(config.datasources[0])
```

**Review by janpio — APPROVED** [review]

File: packages/internals/src/cli/checkUnsupportedDataProxy.ts (hunk 2 above)
Comment (Jolg42) [reply packages/internals/src/cli/checkUnsupportedDataProxy.ts:67]: since it's only a nit Let's merge without this change, once tests are passing

**Comment by Jolg42** [comment]
Sorry for the merge conflicts.... I can check that tomorrow!

**Comment by janpio** [comment]
```
Error: packages/migrate build: src/utils/ensureDatabaseExists.ts(159,31): error TS2304: Cannot find name 'activeDatasource'.
Error: packages/migrate build: src/utils/ensureDatabaseExists.ts(162,49): error TS2345: Argument of type 'string | null' is not assignable to parameter of type 'string'.
packages/migrate build:   Type 'null' is not assignable to type 'string'.
```

File: packages/migrate/src/utils/ensureDatabaseExists.ts [hunk 3]
```diff
@@ -154,9 +156,10 @@ export async function ensureDatabaseExists(action: MigrateAction, schemaPath?: s
   }
 
   const schemaDir = (await getSchemaDir(schemaPath))!
+  const url = getEffectiveUrl(activeDatasource).value
```
Comment (janpio) [inline packages/migrate/src/utils/ensureDatabaseExists.ts:159]:
```suggestion
  const url = getEffectiveUrl(firstDatasource).value
```

File: packages/migrate/src/utils/ensureDatabaseExists.ts [hunk 4]
```diff
@@ -154,9 +156,10 @@ export async function ensureDatabaseExists(action: MigrateAction, schemaPath?: s
   }
 
   const schemaDir = (await getSchemaDir(schemaPath))!
+  const url = getEffectiveUrl(firstDatasource).value
 
-  // url.value exists because `ignoreEnvVarErrors: false` would have thrown an error if not
-  const canConnect = await canConnectToDatabase(firstDatasource.url.value!, schemaDir)
+  // url exists because `ignoreEnvVarErrors: false` would have thrown an error if not
+  const canConnect = await canConnectToDatabase(url, schemaDir)
```
Comment (janpio) [inline packages/migrate/src/utils/ensureDatabaseExists.ts:162]:
```suggestion
  const canConnect = await canConnectToDatabase(url!, schemaDir)
```

**Pull request merged**